### PR TITLE
MAINT:optimize: Comply with user-specified rel_step

### LINKED ISSUE #11258: diff_step in least_squares does absolute step while docs say relative step
The `least_squares` method of `scipy.optimize` has a keyword argument `diff_step`, which allows the user to define the relative step size to be used in computing the numerical Jacobian. The doc strings says: The actual step is computed as ``x * diff_step``. But it, unfortunately, doesn't. It takes an absolute step. In example below the initial parameters are 1000 and 0.01. The `diff_step` is 1e-4, so the Jacobian should be computed with three parameter sets: (1000, 0.01), (1000.1, 0.01), and (1000, 0.010001). But it doesn't do a relative step, it simply adds 1e-4. 

#### Reproducing code example:
```
xobs = np.array([0, 1, 2, 3])
yobs = np.array([0.1, 1005, 1993, 3015])

def residuals(p, xobs=xobs, yobs=yobs):
    print(p)
    return yobs - (p[0] * xobs + p[1])

popt = so.least_squares(residuals, [1000, 0.01], diff_step=1e-4, 
        max_nfev=1)  # just 1 step to show the problem
```

Output (which is incorrect):
```
[1.e+03 1.e-02]
[1.0001e+03 1.0000e-02]
[1.00e+03 1.01e-02]
```

#### Scipy/Numpy/Python version information:
```
1.4.1 1.17.2 sys.version_info(major=3, minor=7, micro=4, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by andyfaff**
In #11258 a user-provided relative step is being used. The relative step is respected if `abs(x0) >= 1`. However if `abs(x0) < 1`, then the user specified relative step is not used because the step is calculated as `rel_step * sign_x0 * np.maximum(1.0, np.abs(x0))`. This PR respects the user provided `rel_step` (even if it's not advisable from a finite different point of view).

pinging @antonior92

Fixes #11258

### CODE REVIEW

**Comment by andyfaff** [comment]
The alternative to this PR is to change the documentation for `least_squares` to reflect what's actually happening.

**Comment by antonior92** [comment]
Hi @andyfaff, thank you for fixing this.

I think you found a good solution: I prefer changing the behavior slightly that updating least-square documentation...

Do you get why it is failing on many CI tests?

**Comment by andyfaff** [comment]
There are two test fails [here](https://travis-ci.org/scipy/scipy/jobs/628541239#L1301), corresponding to https://github.com/scipy/scipy/blob/master/scipy/optimize/tests/test_least_squares.py#L248.

@nmayorov, it's not clear to me why the number of function evaluations needs to be different for `res2` and `res3`: `assert_(res2.nfev != res3.nfev)`. Both solutions successfully find the minimum, albeit with different `diff_step`. What behaviour is that line of the test examining?

If that line is removed the test suite passes.

**Comment by andyfaff** [comment]
ping @nmayorov, what is https://github.com/scipy/scipy/blob/master/scipy/optimize/tests/test_least_squares.py#L246 examining? Why do the number of function evaluations have to be different for the different step sizes used?

**Comment by nmayorov** [comment]
Hello!

1. The last line of the test is not mandatory. I think the logic behind it was to check that changing `diff_step` affects the behavior (and it did). Of course the equal numbers of function evaluations is OK too. So you can easily change it.
2. About the change itself. As far as I understand there is no correct approach for "cheap" numerical differentiation. Some authors recommend to use ``sqrt(eps)`` always, some suggest multiplying by ``x`` (for no strong fundamental reason I'd say). I choose to multiply by ``max(1, x)``,  which also seems reasonable. If we make this change now, I can see that someone else will report an issue, that his step size is getting too small near zero and we should fix that.

As there is no clear advantage of any of these approaches, I'd probably vote for not changing anything. But I'm almost neutral towards that, if someone clearly sees the advantage, please make this change.

By the way, I compared the results om my "lsq-benchmarks" and I didn't notice a single change between master and with PR.

Better explaining in the doc is of course a good idea.

File: scipy/optimize/_numdiff.py [hunk 1]
```diff
@@ -285,18 +303,19 @@ def approx_derivative(fun, x0, method='3-point', rel_step=None, abs_step=None,
...
-        ``h = rel_step * sign(x0) * max(1, abs(x0))``, possibly adjusted to
-        fit into the bounds. For ``method='3-point'`` the sign of `h` is
-        ignored. If None (default) then step is selected automatically,
-        see Notes.
+        Relative step size to use. If None (default) the absolute step size is
+        computed as ``h = rel_step * sign(x0) * max(1, abs(x0))``, with
+        `rel_step` being selected automatically, see Notes. Otherwise
+        ``h = rel_step * sign(x0) * abs(x0)``. For ``method='3-point'`` the
+        sign of `h` is ignored. The calculated step size is possibly adjusted
+        to fit into the bounds.
```
Comment (larsoner) [inline scipy/optimize/_numdiff.py:311]:
```suggestion
        to fit into the bounds, and the automatic step size is always used
        when ``abs(x0) == 0``.
```
Or some other adjustment that makes it clear that it doesn't always use the supplied `rel_step`

**Review by larsoner — COMMENTED** [review]
LGTM other than one minor comment

File: scipy/optimize/_numdiff.py (hunk 1 above)
Comment (andyfaff) [reply scipy/optimize/_numdiff.py:311]: I've improved the notes to make this clearer. `approx_derivative` is still a private function, but it's good to have good docstrings.

**Comment by andyfaff** [comment]
test fail is unrelated.

**Comment by tylerjereddy** [comment]
@nmayorov 

About your comment:

> I'd probably vote for not changing anything

It looks like @larsoner is in favor of the changes--is it possible to reach consensus? I think @antonior92 was also in favor of making the changes.

**Comment by andyfaff** [comment]
The changed behaviour is that user specified relative steps are respected (no longer multiplied by `max(x, 1)`); the behaviour for a default `rel_step = None` is unchanged.
If the user complains that their step is too small near zero, then it's their responsibility, because the code uses the step they specified.

**Comment by tylerjereddy** [comment]
It is labelled "maintenance" instead of bug fix so I'll bump the milestone. If core devs reach a consensus to merge this before I branch 1.6.0 by all means do that and switch the milestone back.

**Comment by tylerjereddy** [comment]
Any chance to reach a consensus here before the 1.7.0 branch cut?

**Comment by andyfaff** [comment]
@tylerjereddy, it seems kind of balanced w.r.t which way at the moment, would be good to get more core-devs to tie-break.

**Comment by antonior92** [comment]
As a way to increase transparency to the user while still avoid letting step size vanish to zero can we add an `abs_step`. And use something like:
```
sign_x0 *( rel_step * np.abs(x0) + abs_step)
```
?

**Comment by ev-br** [comment]
FWIW, I'd rather go with documenting the status quo (or adding `abs_step` like Tyler suggested, with a default value of one).

**Comment by andyfaff** [comment]
I'm coming back to this after a long time away from seeing it.

In the current `least_squares` documentation `diff_step` is said to be a relative step. If `method != 'lm'` then the [Jacobian wrapper](https://github.com/scipy/scipy/blob/v1.6.3/scipy/optimize/_lsq/least_squares.py#L883) around `approx_derivative` will eventually calculate the absolute step as [`rel_step * sign_x0 * np.maximum(1.0, np.abs(x0))`](https://github.com/scipy/scipy/blob/v1.6.3/scipy/optimize/_numdiff.py#L172). This means if the parameter is much less than 1, then the user specified `diffstep` won't be what's actually being used, it'll be a modified version (which is what the original issue is about).

Documenting this is doable, however...

If the least_squares `method == 'lm'` then the `call_minpack` function uses [`epsfcn = diffstep**2`](https://github.com/scipy/scipy/blob/v1.6.3/scipy/optimize/_lsq/least_squares.py#L42) for `_minpack._lmdif`. This is not equivalent to that outlined above. It's unclear to me from the [`lmdif` documentation](https://github.com/scipy/scipy/blob/v1.6.3/scipy/optimize/minpack/lmdif.f#L83) if `epsfcn` is supposed to be the squared relative step.

Thus, with only a documentation change it won't be possible for the docstring to be correct for all methods (unless the docstring for diffstep is expanded to account for what happens with the different methods). If the relative steps are respected for the former case (this PR) then the documentation would be correct as-is. So I don't think that documentation is sufficient.

I don't think @antonior92 suggestion would work because:

1. It would would require significant surgery to `least_squares`.
1. If `a_step` is not specified then then the issues I've just outlined still applies.
2. the equation that Antonio describes results in an absolute step, which can be provided to `approx_derivative`. However, if `method == 'lm'` then approx_derivative is no longer in play, and `lmdif` appears to want a relative step.

**Comment by tylerjereddy** [comment]
I'm still a bit confused on the "majority decision" so I'm bumping the milestone--if it is actually resolved though, feel free to clarify, merge before the branching, and re-milestone.

**Comment by andyfaff** [comment]
I think this is mergeable, I think 2-1 in favour. I'm going to run the tests again, and if they pass I'll merge tomorrow (unless someone pipes up)

**Comment by andyfaff** [comment]
Rebased + force pushed so the proper test suite runs again.

**Comment by mbakker7** [comment]
Thanks for fixing this. Looking forward to trying it!

**Pull request merged**